Movie Recommendation System

Installing necessary libraries

In [111]:
!pip install scikit-learn
!pip install kagglehub
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np
import pandas as pd
import os
import kagglehub
from sklearn.feature_extraction.text import TfidfVectorizer


[notice] A new release of pip is available: 25.0 -> 25.0.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 25.0 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


The data for this project will be the TMDB 5000 movie dataset from Kaggle. To download, run the following code block.

In [112]:
# Download latest version
path = kagglehub.dataset_download("tmdb/tmdb-movie-metadata")

print("Path to dataset files:", path)

Path to dataset files: /Users/arnavpallem/.cache/kagglehub/datasets/tmdb/tmdb-movie-metadata/versions/2


Basic EDA

In [113]:
data = pd.read_csv(os.path.join(path, "tmdb_5000_movies.csv"))
print(data.shape)
data.head()

(4803, 20)


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

We have 4803 movies. The meaningful columns for this task are title and overview for providing some text to match with the user query. 

In [109]:
parsed_data = data[['title', 'overview']].copy()

Now we will vectorize the text

In [110]:
vectorizer = TfidfVectorizer(stop_words = 'english')

tfidf_matrix = vectorizer.fit_transform(parsed_data['overview'].values.astype('str'))
tfidf_vectors = tfidf_matrix.toarray().tolist()
parsed_data.loc[:, 'vectors'] = tfidf_vectors


In [102]:
parsed_data.head() #We now can see a column of vectors

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  vote_average  \
0  In the 22nd century, a paraplegic Marine is di...           7.2   
1  Captain Barbossa, long believed to be dead, ha...           6.9   
2  A cryptic message from Bond’s past sends him o...           6.3   
3  Following the death of District Attorney Harve...           7.6   
4  John Carter is a war-weary, former military ca...           6.1   

                                             vectors  
0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
1  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
2  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
3  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
4  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...

In [114]:
def cs(entry, df_entry):
    return cosine_similarity(entry, np.array(df_entry).reshape(1, -1))[0][0]

def find_similar_movies(entry, vectorizer, df):
    vector = vectorizer.transform([entry])
    temp_df = pd.DataFrame({
        'title': df['title'],
        'similarity': df['vectors'].apply(lambda x: cs(vector, x))
    })
    
    # Sort the temporary DataFrame and return top 5 titles
    return temp_df.sort_values('similarity', ascending=False).head(5)['title']

In [115]:
def communicate():
    while True:
        entry = input("What movie are you looking for?")
        print(find_similar_movies(entry, vectorizer, parsed_data))

Run this block to repeatedly query the system. Press escape to quit

In [116]:
communicate()

618          Mystery Men
2194      Disaster Movie
1375            Sky High
1451                Zoom
403     Last Action Hero
Name: title, dtype: object


KeyboardInterrupt: 